docs/http-cloud-provider.md

```diff
@@ -0,0 +1,643 @@
+<!-- BEGIN MUNGE: UNVERSIONED_WARNING -->
+
+<!-- BEGIN STRIP_FOR_RELEASE -->
+
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+
+<h2>PLEASE NOTE: This document applies to the HEAD of the source tree</h2>
+
+If you are using a released version of Kubernetes, you should
+refer to the docs that go with that version.
+
+<strong>
+The latest 1.0.x release of this document can be found
+[here](http://releases.k8s.io/release-1.0/docs/http-cloud-provider.md).
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# HTTP Cloudprovider
+
+# Motivation
+
+Currently, the cloudproviders are designed so that they get directly compiled into kubernetes. It is very hard to add custom cloudproviders to the kubernetes platform without using a custom build of kubernetes. Therefore, we would like to build an HTTP Cloudprovider which would be able to send a HTTP request by converting the cloudprovider interfaces to an HTTP API. The cloudprovider would need an HTTP endpoint so that kubernetes would be able to send a request to them and then the cloudproviders would be able to send the desired service information back to kubernetes.
+
+# Specification
```

Comment:
Some of the cloudprovider methods access provider zonal, regional, or global infrastructure APIs and therefore require special credentials. They're necessary for the cluster-level control plane, but should not be expected to be callable on arbitrary nodes, or at least not by arbitrary containers. EnsureTCPLoadBalancer is an example.

Others are intended to be accessed from arbitrary nodes, typically using local metadata services. For example, all functions that call metadata.Get in the GCE implementation are in this category. 

There's actually a third category, which is the informational API, such as ProviderName. This should also just be required at the cluster level, I think.

Unfortunately, we haven't clearly distinguished the 3 kinds of functions, and likely are invoking the metadata services on the master node, which is almost certainly wrong. For instance, the master doesn't strictly need to be in the same project, network, or zone as the nodes.

BTW, I believe at least the GCE implementation of the cloudprovider implementation assumes that all nodes are in the same zone. That's not an assumption we want to carry forward.